File: netDxf/IO/DxfLight.cs
// Copyright (c) netDxf contributors. Licensed under the MIT License.
using System;
using System.Collections.Generic;
using System.IO;
using netDxf.Blocks;
using netDxf.Entities;
using netDxf.Header;

namespace netDxf.IO
{
    internal sealed partial class DxfReader
    {
        private Light ReadLight()
        {
            if (this.chunk.Code != 100 || this.chunk.ReadString() != SubclassMarker.Light)
                throw new InvalidDataException("LIGHT requires the AcDbLight subclass.");
            var light = new Light();
            var seen = new HashSet<short>();
            Vector3 position = light.Position, target = light.Target;
            this.chunk.Next();
            while (this.chunk.Code != 0)
            {
                short code = this.chunk.Code;
                if (code == 1001)
                {
                    string app = this.DecodeEncodedNonAsciiCharacters(this.chunk.ReadString());
                    light.XData.Add(this.ReadXDataRecord(this.GetApplicationRegistry(app)));
                    continue;
                }
                // Unknown field preservation is provided by DxfRawDocument, not guessed here.
                if (code == 100 || code == 101 || code == 102)
                    throw new InvalidDataException("Unsupported or duplicate LIGHT subclass/payload marker.");
                bool known = code == 90 || code == 1 || code == 70 || code == 290 || code == 291 || code == 40 ||
                    code == 10 || code == 20 || code == 30 || code == 11 || code == 21 || code == 31 || code == 72 ||
                    code == 292 || code == 41 || code == 42 || code == 50 || code == 51 || code == 293 ||
                    code == 73 || code == 91 || code == 280;
                if (known && !seen.Add(code)) throw new InvalidDataException("Duplicate LIGHT group " + code + ".");
                try
                {
                    switch (code)
                    {
                        case 90: light.VersionNumber = this.chunk.ReadInt(); break;
                        case 1: light.Name = this.DecodeEncodedNonAsciiCharacters(this.chunk.ReadString()); break;
                        case 70: light.LightType = (LightType)this.chunk.ReadShort(); break;
                        case 290: light.IsOn = this.chunk.ReadBool(); break;
                        case 291: light.PlotGlyph = this.chunk.ReadBool(); break;
                        case 40: light.Intensity = this.chunk.ReadDouble(); break;
                        case 10: position.X = this.chunk.ReadDouble(); break;
                        case 20: position.Y = this.chunk.ReadDouble(); break;
                        case 30: position.Z = this.chunk.ReadDouble(); break;
                        case 11: target.X = this.chunk.ReadDouble(); break;
                        case 21: target.Y = this.chunk.ReadDouble(); break;
                        case 31: target.Z = this.chunk.ReadDouble(); break;
                        case 72: light.AttenuationType = (LightAttenuationType)this.chunk.ReadShort(); break;
                        case 292: light.UseAttenuationLimits = this.chunk.ReadBool(); break;
                        case 41: light.AttenuationStartLimit = this.chunk.ReadDouble(); break;
                        case 42: light.AttenuationEndLimit = this.chunk.ReadDouble(); break;
                        case 50: light.HotspotAngle = this.chunk.ReadDouble(); break;
                        case 51: light.FalloffAngle = this.chunk.ReadDouble(); break;
                        case 293: light.CastShadows = this.chunk.ReadBool(); break;
                        case 73: light.ShadowType = (LightShadowType)this.chunk.ReadShort(); break;
                        case 91: light.ShadowMapSize = this.chunk.ReadInt(); break;
                        case 280: light.ShadowMapSoftness = this.chunk.ReadShort(); break;
                        default:
                            if (code >= 1000 && code <= 1071)
                                throw new InvalidDataException("LIGHT XData must start with an application registry.");
                            break;
                    }
                }
                catch (ArgumentException error)
                {
                    throw new InvalidDataException("Invalid LIGHT group " + code + " at position " + this.chunk.CurrentPosition + ".", error);
                }
                this.chunk.Next();
            }
            foreach (short start in new short[] { 10, 11 })
            {
                int count = (seen.Contains(start) ? 1 : 0) + (seen.Contains((short)(start + 10)) ? 1 : 0) +
                    (seen.Contains((short)(start + 20)) ? 1 : 0);
                if (count != 0 && count != 3) throw new InvalidDataException("Incomplete LIGHT point at group " + start + ".");
            }
            light.Position = position; light.Target = target;
            return light;
        }
    }

    internal sealed partial class DxfWriter
    {
        private void ValidateLightVersions()
        {
            if (this.doc.DrawingVariables.AcadVer >= DxfVersion.AutoCad2007) return;
            foreach (Block block in this.doc.Blocks)
                foreach (EntityObject entity in block.Entities)
                    if (entity is Light)
                        throw new NotSupportedException("LIGHT requires AutoCAD 2007 (AC1021) or later in this writer profile.");
        }

        private void WriteLight(Light light)
        {
            this.chunk.Write(100, SubclassMarker.Light);
            this.chunk.Write(90, light.VersionNumber);
            this.chunk.Write(1, this.EncodeNonAsciiCharacters(light.Name));
            this.chunk.Write(70, (short)light.LightType);
            this.chunk.Write(290, light.IsOn); this.chunk.Write(291, light.PlotGlyph);
            this.chunk.Write(40, light.Intensity);
            this.chunk.Write(10, light.Position.X); this.chunk.Write(20, light.Position.Y); this.chunk.Write(30, light.Position.Z);
            this.chunk.Write(11, light.Target.X); this.chunk.Write(21, light.Target.Y); this.chunk.Write(31, light.Target.Z);
            this.chunk.Write(72, (short)light.AttenuationType); this.chunk.Write(292, light.UseAttenuationLimits);
            this.chunk.Write(41, light.AttenuationStartLimit); this.chunk.Write(42, light.AttenuationEndLimit);
            this.chunk.Write(50, light.HotspotAngle); this.chunk.Write(51, light.FalloffAngle);
            this.chunk.Write(293, light.CastShadows); this.chunk.Write(73, (short)light.ShadowType);
            this.chunk.Write(91, light.ShadowMapSize); this.chunk.Write(280, light.ShadowMapSoftness);
            this.WriteXData(light.XData);
        }
    }
}

File: javascript/netDxf/IO/DxfLight.js
import { NullReferenceException } from '../../runtime/Errors.js';
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import { Light } from '../Entities/Light.js';
import { ArgumentException, InvalidDataException, NotSupportedException } from '../../runtime/Errors.js';
import { DecodeDxfText, EncodeDxfDatabaseText } from '../../runtime/DxfStringEncoding.js';
import { ReadXDataRecord, WriteXData, WrappedInvalidData } from '../../runtime/DxfXDataIO.js';
const properties=new Map([[90,['VersionNumber','ReadInt']],[70,['LightType','ReadShort']],[290,['IsOn','ReadBool']],[291,['PlotGlyph','ReadBool']],
  [40,['Intensity','ReadDouble']],[72,['AttenuationType','ReadShort']],[292,['UseAttenuationLimits','ReadBool']],[41,['AttenuationStartLimit','ReadDouble']],
  [42,['AttenuationEndLimit','ReadDouble']],[50,['HotspotAngle','ReadDouble']],[51,['FalloffAngle','ReadDouble']],[293,['CastShadows','ReadBool']],
  [73,['ShadowType','ReadShort']],[91,['ShadowMapSize','ReadInt']],[280,['ShadowMapSoftness','ReadShort']]]);
export function ReadLight(chunk, document) {
  if(chunk.Code!==100 || chunk.ReadString()!=='AcDbLight')throw new InvalidDataException('LIGHT requires the AcDbLight subclass.');
  const light=new Light(),seen=new Set(),position=light.Position,target=light.Target;
  chunk.Next();
  while(chunk.Code!==0) {
    const code=chunk.Code;
    if(code===1001){light.XData.Add(ReadXDataRecord(chunk,document));continue;}
    if(code===100 || code===101 || code===102)throw new InvalidDataException('Unsupported or duplicate LIGHT subclass/payload marker.');
    const scalar=properties.get(code),point=[10,20,30].includes(code)?position:[11,21,31].includes(code)?target:null;
    if(scalar || point || code===1){if(seen.has(code))throw new InvalidDataException('Duplicate LIGHT group '+code+'.');seen.add(code);}
    try {
      if(code===1)light.Name=DecodeDxfText(chunk.ReadString());
      else if(scalar)light[scalar[0]]=chunk[scalar[1]]();
      else if(point)point[code<20?'X':code<30?'Y':'Z']=chunk.ReadDouble();
      else if(code>=1000 && code<=1071)throw new InvalidDataException('LIGHT XData must start with an application registry.');
    } catch(error){if(error instanceof ArgumentException)throw WrappedInvalidData('Invalid LIGHT group '+code+' at position '+chunk.CurrentPosition+'.',error);throw error;}
    chunk.Next();
  }
  for(const start of [10,11]){const count=Number(seen.has(start))+Number(seen.has(start+10))+Number(seen.has(start+20));if(count!==0&&count!==3)throw new InvalidDataException('Incomplete LIGHT point at group '+start+'.');}
  light.Position=position;light.Target=target;return light;
}
export function ValidateLightVersions(document){if(document.DrawingVariables.AcadVer>=15)return;for(const block of document.Blocks)for(const entity of block.Entities)if(entity instanceof Light)throw new NotSupportedException('LIGHT requires AutoCAD 2007 (AC1021) or later in this writer profile.');}
export function WriteLight(chunk, version, light) {
  chunk.Write(100,'AcDbLight');if(light==null)throw new NullReferenceException();chunk.Write(90,light.VersionNumber);chunk.Write(1,EncodeDxfDatabaseText(light.Name,version));
  chunk.Write(70,light.LightType);chunk.Write(290,light.IsOn);chunk.Write(291,light.PlotGlyph);chunk.Write(40,light.Intensity);
  chunk.Write(10,light.Position.X);chunk.Write(20,light.Position.Y);chunk.Write(30,light.Position.Z);
  chunk.Write(11,light.Target.X);chunk.Write(21,light.Target.Y);chunk.Write(31,light.Target.Z);
  chunk.Write(72,light.AttenuationType);chunk.Write(292,light.UseAttenuationLimits);chunk.Write(41,light.AttenuationStartLimit);chunk.Write(42,light.AttenuationEndLimit);
  chunk.Write(50,light.HotspotAngle);chunk.Write(51,light.FalloffAngle);chunk.Write(293,light.CastShadows);chunk.Write(73,light.ShadowType);
  chunk.Write(91,light.ShadowMapSize);chunk.Write(280,light.ShadowMapSoftness);WriteXData(chunk,version,light.XData);
}
